File: Youtube_shots_Maker/backend/models/blog.py
"""
models/blog.py
Track D: 네이버 블로그 포스트 — Pydantic 스키마 (Phase 4)
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel


# ─────────────────────────────────────────────────────────────
# 요청 스키마
# ─────────────────────────────────────────────────────────────

class BlogGenerateRequest(BaseModel):
    video_id: str
    script_id: Optional[str] = None          # 없으면 해당 video_id의 최신 스크립트 사용
    style: Literal["review", "info", "story"] = "review"
    commerce_link: Optional[str] = None      # 생성 시 {{COMMERCE_LINK}} 교체


class BlogUpdateRequest(BaseModel):
    title_ko: Optional[str] = None
    body_html: Optional[str] = None
    tags: Optional[list[str]] = None
    commerce_links: Optional[dict] = None    # { coupang?, smartstore?, naver_shopping? }
    publish_account_id: Optional[str] = None


class BlogPublishRequest(BaseModel):
    blog_post_id: str
    account_id: str                          # naver_blog 계정 UUID
    scheduled_at: Optional[datetime] = None
    proxy_id: Optional[str] = None


# ─────────────────────────────────────────────────────────────
# 응답 스키마
# ─────────────────────────────────────────────────────────────

class BlogTaskResponse(BaseModel):
    task_id: str
    blog_post_id: str
    status: str
    message: str


class BlogPostResponse(BaseModel):
    id: str
    user_id: str
    video_id: Optional[str] = None
    script_id: Optional[str] = None
    publish_account_id: Optional[str] = None
    title_ko: str
    body_html: str
    thumbnail_path: Optional[str] = None
    tags: Optional[list[str]] = None
    commerce_links: Optional[dict] = None
    naver_post_url: Optional[str] = None
    status: str
    generate_task_id: Optional[str] = None
    publish_task_id: Optional[str] = None
    error_msg: Optional[str] = None
    published_at: Optional[datetime] = None
    created_at: datetime
    # 계정 정보 (JOIN)
    account_alias: Optional[str] = None
    account_display_name: Optional[str] = None


class BlogJobItem(BaseModel):
    """배포 작업 목록용 경량 응답"""
    blog_post_id: str
    title_ko: str
    status: str
    naver_post_url: Optional[str] = None
    published_at: Optional[datetime] = None
    error_msg: Optional[str] = None
    account_alias: Optional[str] = None
    account_display_name: Optional[str] = None


class BlogTaskStatusResponse(BaseModel):
    task_id: str
    blog_post_id: Optional[str] = None
    status: str
    progress: int
    result: Optional[dict] = None
    error: Optional[str] = None

File: Youtube_shots_Maker/lib/api/blog.ts
import { api } from "./client";

// ─── 타입 ───────────────────────────────────────────────────

export type BlogStyle = "review" | "info" | "story";

export type BlogStatus =
  | "draft"
  | "generating"
  | "ready"
  | "publishing"
  | "published"
  | "failed";

export interface CommerceLinks {
  coupang?: string;
  smartstore?: string;
  naver_shopping?: string;
}

export interface BlogPost {
  id: string;
  user_id: string;
  video_id: string | null;
  script_id: string | null;
  publish_account_id: string | null;
  title_ko: string;
  body_html: string;
  thumbnail_path: string | null;
  tags: string[] | null;
  commerce_links: CommerceLinks | null;
  naver_post_url: string | null;
  status: BlogStatus;
  generate_task_id: string | null;
  publish_task_id: string | null;
  error_msg: string | null;
  published_at: string | null;
  created_at: string;
  account_alias: string | null;
  account_display_name: string | null;
}

export interface BlogJobItem {
  blog_post_id: string;
  title_ko: string;
  status: BlogStatus;
  naver_post_url: string | null;
  published_at: string | null;
  error_msg: string | null;
  account_alias: string | null;
  account_display_name: string | null;
}

export interface BlogTaskResponse {
  task_id: string;
  blog_post_id: string;
  status: string;
  message: string;
}

export interface BlogTaskStatus {
  task_id: string;
  blog_post_id: string | null;
  status: string;
  progress: number;
  result?: Record<string, unknown>;
  error?: string;
}

export interface GenerateRequest {
  video_id: string;
  script_id?: string;
  style?: BlogStyle;
  commerce_link?: string;
}

export interface PublishRequest {
  blog_post_id: string;
  account_id: string;
  scheduled_at?: string;
  proxy_id?: string;
}

export interface UpdateRequest {
  title_ko?: string;
  body_html?: string;
  tags?: string[];
  commerce_links?: CommerceLinks;
  publish_account_id?: string;
}

// ─── API 함수 ────────────────────────────────────────────────

export const blogApi = {
  generate: (body: GenerateRequest) =>
    api.post<BlogTaskResponse>("/api/blog/generate", body),

  listPosts: (params?: { video_id?: string; status?: BlogStatus }) => {
    const qs = new URLSearchParams(
      Object.fromEntries(Object.entries(params ?? {}).filter(([, v]) => v))
    ).toString();
    return api.get<BlogPost[]>(`/api/blog/posts${qs ? `?${qs}` : ""}`);
  },

  preview: (blog_post_id: string) =>
    api.get<BlogPost>(`/api/blog/preview/${blog_post_id}`),

  update: (blog_post_id: string, body: UpdateRequest) =>
    api.put<BlogPost>(`/api/blog/${blog_post_id}`, body),

  publish: (body: PublishRequest) =>
    api.post<BlogTaskResponse>("/api/blog/publish", body),

  listJobs: (account_id?: string) => {
    const qs = account_id ? `?account_id=${account_id}` : "";
    return api.get<BlogJobItem[]>(`/api/blog/jobs${qs}`);
  },

  taskStatus: (task_id: string) =>
    api.get<BlogTaskStatus>(`/api/blog/task/${task_id}`),
};
